src/cli/commands/test/mod.rs

```diff
@@ -112,22 +151,120 @@ fn list_cairo_files(root: &PathBuf) -> Result<Vec<PathBuf>, String> {
+				return Some((cache.path, compiled_path, entrypoints));
 			},
 		},
-		Err(e) => {
-			eprintln!("{}", e);
-			None
+		Err(err) => {
+			eprintln!("{}", err);
+			return None;
+		},
+	}
+}
+
+fn create_compiled_contract_path(path_to_code: &PathBuf) -> PathBuf {
+	let filename = path_to_code.file_stem().expect("File does not have a file stem");
+
+	let cache_dir = dirs::cache_dir().expect("Could not make cache directory");
+	let mut path_to_compiled = PathBuf::new();
+	path_to_compiled.push(&cache_dir);
+	path_to_compiled.push("compiled-cairo-files");
+	path_to_compiled.push(filename);
+	path_to_compiled.set_extension("json");
+	return path_to_compiled;
+}
+
+fn dump_json_file(path: &PathBuf, data: &CacheJson) -> Result<(), String> {
+	let file = File::create(path).map_err(|op| format!("file does not exists {}", op))?;
+	let writer = BufWriter::new(file);
+	serde_json::to_writer_pretty(writer, data)
+		.map_err(|op| format!("file does not exists {}", op))?;
+	return Ok(());
+}
+
+fn read_cache(path_to_code: PathBuf) -> Result<CompiledCacheFile, String> {
+	// read individual cache file
+	// avoid same cache file because we're doing multiprocessing and getting race condition
+	let cache_dir = dirs::cache_dir().expect("cache dir not supported");
+	let filename = path_to_code.file_stem().unwrap().to_str().unwrap();
+
+	let mut cache_path = PathBuf::new();
+	cache_path.push(&cache_dir);
+	cache_path.push("cairo-foundry-cache");
+
+	// create dir if not exist to store cache files
+	// cache dir will be in os_cache_dir/cairo-foundry-cache
+	// os_cache_dir is different for each os
+	if !cache_path.exists() {
+		std::fs::create_dir(&cache_path).expect("Could not make cache directory");
+	}
+	// cache file will be in os_cache_dir/cairo-foundry-cache/contract_name.json
+	cache_path.push(format!("{}.json", filename));
+
+	let data = read_json_file(&cache_path);
+	// compute hash from file
+	let hash_calculated = compute_hash(&path_to_code).unwrap();
+	let contract_path = path_to_code.to_str().unwrap().to_string();
+
+	match data {
+		// json file exists
+		Ok(cache_data) => {
+			let compiled_contract_path = create_compiled_contract_path(&path_to_code);
+			let hash_in_cache = cache_data.sha256;
+			if *hash_in_cache == hash_calculated {
+				return Ok(CompiledCacheFile {
+					path: compiled_contract_path,
+					status: CacheStatus::Cached,
+				});
+			} else {
+				let data = CacheJson {
+					contract_path,
+					sha256: hash_calculated,
+				};
+
+				dump_json_file(&cache_path, &data)?;
+				return Ok(CompiledCacheFile {
+					path: path_to_code,
+					status: CacheStatus::Uncached,
+				});
+			}
+		},
+
+		// json file does not exists
+		Err(_) => {
+			let data = CacheJson {
+				contract_path,
+				sha256: hash_calculated,
+			};
+			dump_json_file(&cache_path, &data)?;
+			return Ok(CompiledCacheFile {
+				path: path_to_code,
+				status: CacheStatus::Uncached,
+			});
```

Comment:
Return a meaningful error and only create the `.json` cached file later on, after you compiled it